File: Sources/parser/expression.c
#include <stdbool.h>

struct expression;
typedef struct expression expression_t;

struct MemberExpression {
	const char *value1;
	const char *value2;
};

struct ExprBinary {
	expression_t *left;
	Operator op;
	expression_t *right;
};

struct ExprUnary {
	Operator op;
	expression_t *right;
};

struct ConstructorExpression {
	Vec<Expression> parameters;
};

struct expression_t {
	enum { Binary, Unary, Boolean, Number, String, Variable, Grouping, Call, Member, Constructor } type;

	union {
		ExprBinary binary;
		ExprUnary unary;
		bool boolean;
		float number;
		const char *string;
		const char *variable;
		expression_t *grouping;
		CallStatement call;
		MemberExpression member;
		ConstructorExpression constructor;
	} data;
};

File: Sources/parser/expression.rs
use super::statement::*;
use super::super::tokenizer::token::*;

pub enum Expression {
	Binary(ExprBinary),
	Unary(ExprUnary),
	Boolean(bool),
	Number(f64),
	Stringy(String),
	Variable(String),
	Grouping(Box<Expression>),
	Call(CallStatement),
	Member(MemberExpression),
	Constructor(ConstructorExpression),
}

pub struct MemberExpression {
	pub value1: String,
	pub value2: String,
}

pub struct ExprBinary {
	pub left: Box<Expression>,
	pub op: Operator,
	pub right: Box<Expression>,
}

pub struct ExprUnary {
	pub op: Operator,
	pub right: Box<Expression>,
}

pub struct ConstructorExpression {
	pub parameters: Vec<Expression>,
}
